### DISCUSSION TITLE
Live plotting feature

### DISCUSSION THREAD

**Discussion by Sreevis** [alandefreitas/matplotplusplus#124] [Q&A] [answered]
Hi,

Does this library provide any live plotting functionality? 

I was trying to integrate plotting libraries with cling/Jupyter for doing some live plotting. Was able to integrate matplotplusplus with jupyter. Static plots are working fine. But I need to do some live plotting.

Please give some advice if anybody has worked on it.

Thanks,
Sreevishnu

**Comment by alandefreitas** [comment, marked as answer]
If you mean plotting inside a loop or something like that, there should be no problem. Lots of examples do that. 

I don't see any reason why it shouldn't work on xeus-cling either. I often use xeus-cling / jupyter / matplot++ for short experiments, etc.

The only problem here could be the performance because gnuplot pipes are a little slow if you need many frames per second, and the OpenGL canvas is still on the way. But it should be fine for most applications.

**Reply by Sreevis** [reply]
Thanks for the quick reply.

Happy to see that it is possible. Can we embed the plots inside the notebook? When I tried it was running separately. But it is fine for my requirement. 

It would be great if you can share some sample notebooks.

What I am exactly trying to do is live control of plot using [xwidgets(slider).](https://github.com/jupyter-xeus/xwidgets).

Thanks

**Reply by alandefreitas** [reply]
You *could* save the image and write a function to open and display it directly in the notebook with something like

```cpp
auto bundle = nl::json::object();
bundle["image/png"] = xtl::base64encode(your_image);
return bundle;
```

using the [image functions](https://alandefreitas.github.io/matplotplusplus/plot-types/images/image-show/) to create `your_image`. You could also set the Gnuplot terminal so that the window doesn't show up when you are creating the plots.

But that's probably a bad idea at the moment because it's not going to go around Gnuplot anyway. So it's more convenient to just look at the plots in the window for now. 

You can use the widgets either way.

That's not to say that plots directly in the notebook would not be convenient very often. Especially for scientific computing. I do have plans to create a canvas / backend that would directly put the plots in the notebook, but doing that before the OpenGL backend is ready is kind of a workaround.

I've been using matplot++ with notebooks quite a lot lately. But I'm not sure someone would let me share the notebooks (for free 😆) unless I run my own server because I would need them to let me install matplot++ as a dependency. 

Maybe there's a way to do that. I don't have a lot of experience with notebooks. I'm just a regular user.

**Reply by Sreevis** [reply]
Thank you so much @alandefreitas. 
Tried ur suggestions and I am getting the required output.
